plugins/@grouparoo/pardot/src/lib/client.ts

```diff
@@ -0,0 +1,215 @@
+  id: number;
+  name: string;
+  fieldId: string;
+  type: string;
+  isRecordMultipleResponses: boolean;
+}
+
+class PardotClient {
+  private pardotUrl: string;
+  private businessUnitId: string;
+
+  conn: jsforce.Connection;
+  request?: AxiosInstance;
+
+  constructor(options?: PardotClientOptions) {
+    this.pardotUrl = options?.pardotUrl || "https://pi.pardot.com";
+    this.businessUnitId = options.businessUnitId;
+
+    this.conn = new jsforce.Connection({
+      loginUrl: options?.loginUrl || "https://login.salesforce.com",
+    });
+  }
+
+  async login(username: string, password: string, securityToken?: string) {
+    const fullPassword =
+      (password || "").toString() + (securityToken || "").toString();
+
+    await this.conn.login(username, fullPassword);
+
+    this.request = axios.create({
+      baseURL: `${this.pardotUrl}/api`,
+      headers: {
+        Authorization: `Bearer ${this.conn.accessToken}`,
+        "Pardot-Business-Unit-Id": this.businessUnitId,
+      },
+    });
+  }
+
+  async getCustomFields(): Promise<PardotCustomField[]> {
+    const res = await this.request.get("/customField/version/4/do/query", {
+      params: {
+        format: "json",
+        output: "mobile", // mobile output does not require pagination
+      },
+    });
+
+    const { result } = res.data;
+
+    const mapResult = (field: any): PardotCustomField => ({
+      id: field.id,
+      name: field.name,
+      fieldId: field.field_id,
+      type: field.type,
+      isRecordMultipleResponses: field.is_record_multiple_responses,
+    });
+
+    if (result.total_results === 0) {
+      return [];
+    } else if (result.total_results === 1) {
+      return [mapResult(result.customField)];
+    }
+
+    return result.customField.map(mapResult);
+  }
+
+  async getProspectByEmail(email: string) {
+    try {
+      const res = await this.request.get(
+        `/prospect/version/4/do/read/email/${email}`,
+        {
+          params: {
+            format: "json",
+          },
+        }
+      );
+
+      if (Array.isArray(res.data.prospect)) {
+        return res.data.prospect[0];
+      }
+
+      return res.data.prospect;
+    } catch (error) {
+      if (
+        error.response?.status === 400 &&
+        error.response?.data["@attributes"]?.err_code === 4
+      ) {
+        return null;
+      }
+
+      throw error;
+    }
+  }
+
+  async getProspectById(id: number) {
+    try {
+      const res = await this.request.get(
+        `/prospect/version/4/do/read/id/${id}`,
+        {
+          params: {
+            format: "json",
+          },
+        }
+      );
+
+      if (Array.isArray(res.data.prospect)) {
+        return res.data.prospect[0];
+      }
+
+      return res.data.prospect;
+    } catch (error) {
+      if (
+        error.response?.status === 400 &&
+        error.response?.data["@attributes"]?.err_code === 4
+      ) {
+        return null;
+      }
+
+      throw error;
+    }
+  }
+
+  async queryProspects(
```

Comment:
this is paginated right? don't want to blow out the memory.

Reply:
Yeah, but the `TestPluginMethod` was overriding it with `output=mobile`. Changed this so it respects the pagination when testing connection (93bd712)